### PR TITLE
fix: always ensure endpoint creation

### PR DESCRIPTION
**Pull request by derekbit**
#### Which issue(s) this PR fixes:

Issue Longhorn/longhorn#9919

- fix: always ensure endpoint creation
- fix: reload device numbers while reusing linear dm device
- refactor: improve log messages and api docs

Signed-off-by: Derek Su <derek.su@suse.com>

### CODE REVIEW

File: pkg/util/device.go [hunk 1]
```diff
@@ -41,13 +41,15 @@ type LonghornBlockDevice struct {
...
+func RemoveDevice(devPath string) error {
+	if _, err := os.Stat(devPath); err != nil {
+		if os.IsNotExist(err) {
+			return nil
 		}
 	}
-	return nil
+
+	// Still try to remove the device node
+	return remove(devPath)
```
[the remark this answers is not in the document]
Comment (derekbit) [reply pkg/util/device.go:52]: No, although `stat` returns error except for `os.IsNotExist(err)`, we still try to remove the device. It is a best-effort deletion.

File: pkg/nvme/nvmecli.go [hunk 2]
```diff
@@ -149,12 +149,12 @@
 	}
 
 	if major == 1 {
-		return listSubsystemsV1(jsonStr, executor)
+		return listSubsystemsV1(jsonStr)
 	}
-	return listSubsystemsV2(jsonStr, executor)
+	return listSubsystemsV2(jsonStr)
```
[the remark this answers is not in the document]
Comment (derekbit) [reply pkg/nvme/nvmecli.go:154]: `executor` is not used by listSubsystemsV1 and listSubsystemsV2, so no need to pass it to the functions.

File: pkg/nvme/initiator.go [hunk 3]
```diff
@@ -271,104 +270,113 @@
...
+	for r := 0; r < maxNumWaitDeviceRetries; r++ {
+		//err = i.loadNVMeDeviceInfoWithoutLock(i.TransportAddress, i.TransportServiceID, i.SubsystemNQN)
+		err = i.loadNVMeDeviceInfoWithoutLock(transportAddress, transportServiceID, i.SubsystemNQN)
+		if err == nil {
+			break
+		}
+		time.Sleep(waitDeviceInterval)
+	}
+	return err
+}
```
[the remark this answers is not in the document]
Comment (derekbit) [reply pkg/nvme/initiator.go:354]: They are defined in another go file.

File: pkg/nvme/initiator.go [hunk 4]
```diff
@@ -270,105 +271,113 @@
...
+		return dmDeviceIsBusy, errors.Wrapf(err, "failed to check if endpoint %v exists for NVMe-oF initiator %s", i.Endpoint, i.Name)
+	}
+	if exist {
+		i.logger.Infof("Skipping endpoint %v creation for NVMe-oF initiator", i.Endpoint)
+	} else {
 		if err := i.makeEndpoint(); err != nil {
-			return dmDeviceBusy, err
+			return dmDeviceIsBusy, err
 		}
 	}
```
[the remark this answers is not in the document]
Comment (derekbit) [reply pkg/nvme/initiator.go:339]: We will have a lock in the caller side, so no need to add a lock here.

**Review by c3y1huang — APPROVED** [review]

**Pull request merged**